packages/nimble-components/src/table/specs/table-row-grouping-hld.md

```diff
@@ -0,0 +1,238 @@
+# Row Grouping HLD
+
+## Problem Statement
+
+There is a requirement that users be able to group together rows of data under collapsible sections based on unique values in a particular column. Users need to be able to group by values in more than one column which will result in a collapsible "header" row for each grouped value.
+
+This document will focus on the programmatic grouping of data rows.
+
+### Out of Scope of this HLD
+
+1. The proposal in this HLD is focusing on the broad strategy for providing a means for columns to specify how to render the group header associated with that column's values. It will not focus on particular concerns that some columns may have for how to render their value, such as any localization requirements. Ultimately, this strategy should provide a means to those ends without getting into the details of each.
+2. This proposal will not provide a means for rendering values from multiple columns for the group header. TanStack does not support this, and we will not block ourselves on this restriction.
+3. Interactive grouping concerns.
+
+## Links To Relevant Work Items and Reference Material
+
+-   [#1003 Programmatic grouping of data rows](https://github.com/ni/nimble/issues/1003)
+-   [IxD Document](https://xd.adobe.com/view/6f3be15d-8503-4f1f-54b9-5bc27955b3e4-190a/screen/13edcacf-4e95-46b2-a7a8-83141afb3f2d)
+-   [Table README](./README.md)
+-   [Table Design Doc](https://xd.adobe.com/view/5b476816-dad1-4671-b20a-efe796631c72-0e14/screen/d389dc1e-da4f-4a63-957b-f8b3cc9591b4/specs/)
+-   [Prototype branch](https://github.com/ni/nimble/tree/table-row-grouping-prototype) ([Storybook](https://60e89457a987cf003efc0a5b-bwzrahvxqm.chromatic.com/?path=/story/table--table))
+
+## Implementation / Design
+
+![Row Grouping Example](./spec-images/row-groups.PNG)
+
+### Row Grouping Column API
+
+To support both the notion that a column type may want to allow grouping in some situations and not others (e.g. a numeric column displaying discrete values versus continuous values), as well as the possibility that certain columns should never be groupable, we will create a mixin to provide the necessary grouping APIs:
+
+```
+export function mixinGroupableColumnAPI<
+    TBase extends GroupableTableColumnConstructor
+>(base: TBase) {
+    abstract class GroupableColumn extends base {
+        public disableGrouping?: boolean;
```

Comment:
My primary reservation is that I believe the guidance is that if an attribute is of type `boolean` its default value should be `false`. @rajsite or @mollykreis , do you have any preference here?

Reply:
So, the case that I don't think is handled in the above scenario is for providing state that a user can use to determine whether or not they should allow a column (whose type allows grouping...that is, implements the `GroupableColumn` mixin), to be able to be grouped. Take a numeric column, which could ultimately support a set of discrete values or continuous values, but the client wants to prevent their users from grouping by the numeric column that displays continuous values. This would have to be provided by state beyond `groupIndex`.

Reply:
The concern was expressing these two concepts:

> can this column ever be grouped? (i.e. is its "group by" button in the header enabled?)
is this column currently being grouped?

What I was proposing in a round about way is that "is this column currently being grouped" is already expressed by group index if we use the column width pattern (via currentGroupIndex).

So:
- can this column ever be grouped (user can use to determine whether or not they should allow a column)? -> disableGrouping (which the mixin maps to internalIsGroupable)
- is this column currently being grouped? -> currentGroupIndex